chore: create spike template (#3245)
statechannels/statechannels: .github/ISSUE_TEMPLATE/spike.md

---
name: Spike
about: Research and/or Analysis Required
title: ''
labels: ":zap: spike"
assignees: ''

---

# Why: Problem description :grey_question:
A clear and concise description of what the problem is. 
- [ ] Describe the matter that needs investigating
- [ ] Provide context for posterity

# What: Describe what a solution should resolve :grey_exclamation:  

> A clear and concise description of what a working solution should resolve, and any required deliverables (i.e. issue cards, prototypes, POCs etc.)

Reviewer: Let's take an example, the challenge protocol v0

The intention here is to elicit the issue author to write something like:
> The expected sequence of events is roughly
> 1. blah
> 2. foo
> 3. fizzbuzz
> This spike should generate cards to implement this story.

right?

Author: This is more to define the "exit criteria" for a spike. It is not exclusive to breaking down a project into issues.
Spikes are often used to refine lower level criteria or determine implementation approach, which can lead to the creation of more cards, refining existing cards, and sometimes new epics.

For example:
In the case of choosing a ledger funding approach, "Analyze both ledger funding solutions, discuss recommendation with x and y, update notion documentation with the agreed upon solution (alternatively, create the issue cards to implement the chosen approach)"
"Provide an analysis of various 3rd party APIs to retrieve user location data. This 3rd party API should be able to define user location based on browser data, ip address, or user provided postal code."

# Describe any ideas you've considered :bulb:

> A clear and concise description of any alternative solutions or features you've considered.
> Implementation suggestions or options may be provided.

Reviewer: ```suggestion
A clear and concise description of the solutions or features you've considered.
Implementation suggestions or options may be provided.
```

**Proposed timebox**
(Optional) in days

**Additional context**
Add any other context or screenshots about the feature request here: notion doc, github discussion or ticket

# Project management
- [ ] I have created any required issues or epics

> - [ ] I have updated the related notion or epic documentation

Reviewer: Are these intended to be checked off as the spike is being completed? If so, it might be worth adding a note to clarify this -- another interpretation is "you should check these off as you are filling out this template".

Author: It was meant to mimic your pull request template, like a reminder to update issues / epic / notion documentation.

Reviewer: Right. My question is:

1. Am I intended to check them off _before submitting the issue_?
2. Or am I intended to check them off _as I "execute" the spike_, to indicate that I'm done with it?

I think it's (2), so maybe this merits a bit of clarification, such as:
```suggestion
Before closing this card,
- [ ] I have created any required issues or epics
- [ ] I have updated the related notion or epic documentation
```

Author: Sounds good!

- [ ] Provide context for posterity

# What: Describe what a solution should resolve :grey_exclamation:  

> Define the 'exit criteria' for the spike, e.g. whether a required deliverable is to be created (i.e. issue cards, prototypes, POCs) or an implementation decision should be documented (i.e. notion or refining issue cards).

Reviewer: :+1:
